src/react/providers/chat-room-provider.tsx

```diff
@@ -77,42 +128,116 @@ export const ChatRoomProvider: React.FC<ChatRoomProviderProps> = ({
-        return prev;
+  useEffect(() => {
+    logger.debug(`ChatRoomProvider(); running room options change effect`, { roomId, options });
+    if (client === prevClient.current && prevId.current === roomId && prevOptions.current === options) {
+      return;
+    }
+
+    prevId.current = roomId;
+    prevOptions.current = options;
+    prevClient.current = client;
+
+    // The room options or id is changing in some way, so we need to release the old room
+    // before we create a new one
+    setValue((prev: ChatRoomContextType) => {
+      logger.debug(`ChatRoomProvider(); running options change setValue`, { roomId, options });
+      // If we're releasing, release the room - it's guaranteed to be changing
+      if (release) {
+        logger.debug(`ChatRoomProvider(); releasing value`, { roomId });
+        void client.rooms.release(prev.roomId).catch();
+      } else if (attach) {
+        // If we're not releasing, but we're attaching, detach the room
+        void prev.room.then((room) => {
+          logger.debug(`ChatRoomProvider(); detaching value`, { roomId });
+          void room.detach().catch();
+        });
       }
-      return { room };
+
+      logger.debug(`ChatRoomProvider(); updating value`, { roomId, options });
+      return { room: client.rooms.get(roomId, options).catch(), roomId, options };
     });
+  }, [client, roomId, options, logger, attach, release]);
 
-    if (attach) {
-      // attachment error and/or room status is available via useRoom
-      // or room.status, no need to do anything with the promise here
-      logger.debug(`ChatRoomProvider(); attaching room`, { roomId });
-      void room.attach().catch(() => {
-        // Ignore, the error will be available via various room status properties
-      });
-    }
+  // Create an effect that attaches and detaches the room, or releases it when the component unmounts
+  useEffect(() => {
+    let unmounted = false;
+    let resolvedRoom: Room | undefined;
+    const currentReleaseQueue = roomReleaseQueue.current;
+    logger.debug(`ChatRoomProvider(); running attachment and release effect`, { roomId: value.roomId });
+
+    // If there was a previous release queued for this room, abort it
+    currentReleaseQueue.abort(value.roomId, value.options);
+
+    // Get the latest value of the room promise to ensure we're not using stale values
+    void value.room
+      .then((room: Room) => {
+        if (unmounted) {
+          logger.debug(`ChatRoomProvider(); unmounted before room resolved`, { roomId: value.roomId });
+        }
+
+        logger.debug(`ChatRoomProvider(); room resolved`, { roomId: value.roomId });
+
+        resolvedRoom = room;
+
+        if (attach) {
+          // attachment error and/or room status is available via useRoom
+          // or room.status, no need to do anything with the promise here
+          logger.debug(`ChatRoomProvider(); attaching room`, { roomId: value.roomId });
+          void room.attach().catch(() => {
```

Comment:
Is the void needed now?

Reply:
It isn't, but I think it makes it very clear that we're not monitoring the promise result :)